pandas/core/categorical.py

```diff
@@ -1660,16 +1660,26 @@ def fillna(self, value=None, method=None, limit=None):
 
         else:
 
-            if not isna(value) and value not in self.categories:
-                raise ValueError("fill value must be in categories")
+            if isinstance(value, ABCSeries):
+                if not value[~value.isin(self.categories)].isna().all():
+                    raise ValueError("fill value must be in categories")
 
-            mask = values == -1
-            if mask.any():
-                values = values.copy()
-                if isna(value):
-                    values[mask] = -1
-                else:
-                    values[mask] = self.categories.get_loc(value)
+                values_codes = _get_codes_for_values(value, self.categories)
+                indexer = np.where(values_codes != -1)
+                values[indexer] = values_codes[values_codes != -1]
+
+            # Scalar value
+            else:
+                if not isna(value) and value not in self.categories:
+                    raise ValueError("fill value must be in categories")
+
+                mask = values == -1
```

Comment:
you might be able to share code with the array case (basically convert the scalar to an array)

Reply:
Yeah, I was thinking about that but couldn't get it to work properly. The problem I was having was that if the user passes a scalar all the NaNs will be filled with that single scalar. But if the user passes a Series (or a dict which is then converted to a Series in the function) the NaNs will be filled with different values according to the index values:
```
In [1]: data = ['a', np.nan, 'b', np.nan, np.nan]
In [2]: s = pd.Series(pd.Categorical(data, categories=['a', 'b']))

In [3]: s.fillna('a')
Out[3]: 
0    a
1    a
2    b
3    a
4    a
dtype: category
Categories (2, object): [a, b]

In [4]: s.fillna(pd.Series(['a', 'b', 'b'], index=[1, 3, 4])
Out[4]:
0    a
1    a
2    b
3    b
4    b
dtype: category
Categories (2, object): [a, b]
```
And I was finding it difficult to deal with both cases with the same code. But I can keep thinking about it.